File: 07.bridge/bridge.js
// Implementación
class Platform {
    execute(command) {
        throw new Error("Método no implementado");
    }
}

// Implementaciones concretas
class WindowsPlatform extends Platform {
    execute(command) {
        console.log(`Ejecutando en Windows: ${command}`);
    }
}

class LinuxPlatform extends Platform {
    execute(command) {
        console.log(`Ejecutando en Linux: ${command}`);
    }
}

// Abstracción
class SmartDevice {
    constructor(platform) {
        this.platform = platform;
    }

    performTask(task) {
        throw new Error("Método no implementado");
    }
}

// Abstracciones concretas
class RemoteControl extends SmartDevice {
    performTask(task) {
        this.platform.execute(`Control remoto: ${task}`);
    }
}

class Statistics extends SmartDevice {
    performTask(task) {
        this.platform.execute(`Estadísticas: ${task}`);
    }
}

// Uso
const remote = new RemoteControl(new WindowsPlatform());
remote.performTask("Encender TV");

const stats = new Statistics(new LinuxPlatform());
stats.performTask("Recopilar datos");

File: 07.bridge/bridge.py
from abc import ABC, abstractmethod

# Implementación
class Platform(ABC):
    @abstractmethod
    def execute(self, command):
        pass

# Implementaciones concretas
class WindowsPlatform(Platform):
    def execute(self, command):
        print(f"Ejecutando en Windows: {command}")

class LinuxPlatform(Platform):
    def execute(self, command):
        print(f"Ejecutando en Linux: {command}")

# Abstracción
class SmartDevice(ABC):
    def __init__(self, platform: Platform):
        self.platform = platform

    @abstractmethod
    def perform_task(self, task):
        pass

# Abstracciones concretas
class RemoteControl(SmartDevice):
    def perform_task(self, task):
        self.platform.execute(f"Control remoto: {task}")

class Statistics(SmartDevice):
    def perform_task(self, task):
        self.platform.execute(f"Estadísticas: {task}")

# Uso
remote = RemoteControl(WindowsPlatform())
remote.perform_task("Encender TV")

stats = Statistics(LinuxPlatform())
stats.perform_task("Recopilar datos")
